feat: add `math/base/special/exp2f`

aayush0325 opened the pull request:
Progresses #649

## Description

> What is the purpose of this pull request?

This pull request:

-   adds `math/base/special/exp2f`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #649

## Questions

> Any questions for reviewers of this pull request?

What should i do about `benchmark/c/cephes` ? that doesn't seem to compile locally and on the CI tests as well.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Commit:
feat: basic stuff added

Commit:
feat: auto generated functions

Commit:
feat: matching tolerance for tests

Commit:
feat: some changes

Commit:
feat: examples added

Commit:
feat: docs added

Commit:
feat: benchmarks added

aayush0325 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Commit:
fix: removed bugs

Commit:
chore: removing cephes benchmarks for now

Commit:
chore: spacing error

Planeshifter reviewed (changes requested):
Thanks for your work on this PR! Looks great overall, but some changes will be needed for the JavaScript implementation.

In `main.js`, we need to use the helper function

```js
var float64ToFloat32 = require( '@stdlib/number/float64/base/to-float32' );
```

in order to ensure that all operations are done in a manner consistent with single-precision arithmetic. Recall that in JavaScript, all numbers are double-precision by default, so to emulate single-precision we must, in addition to using the single-precision variants of math functions, explicitly cast intermediate results and constants to single-precision floats.

aayush0325 committed:
feat: explicitly convert values to float32

aayush0325 commented:
good morning @Planeshifter! I've wrapped the intermediate results and final answers in `float64ToFloat32()`. If there's anything else that needs to be changed here please let me know.

aayush0325 commented:
Also, as mentioned in the questions section of the PR, what should i do about `benchmark/c/cephes` since it doesn't seem to compile locally and it fails on CI as well. Is there something that I'm missing from my side? (see [this commit](https://github.com/stdlib-js/stdlib/pull/3366/commits/1ab71c012c90f80851c4f3d10ba3e94c6eef74c5))

gunjjoshi commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/package.json:
```diff
@@ -0,0 +1,72 @@
+    "exp",
+    "exp2",
+    "base 2",
+    "natural",
```
```suggestion
```

gunjjoshi replied:
"natural" is for base-e, i.e., `exp`. But, `exp2f` is base-2.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/package.json:
```diff
@@ -0,0 +1,72 @@
+    "base 2",
+    "natural",
+    "exponential",
+    "euler",
```
```suggestion
```

gunjjoshi replied:
For the same reason as above, we can remove this too, as this is base 2, not base `e`.

gunjjoshi committed:
Apply suggestions from code review

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/README.md:
````diff
@@ -0,0 +1,188 @@
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var randu = require( '@stdlib/random/base/randu' );
+var exp2f = require( '@stdlib/math/base/special/exp2f' );
+
+var x;
+var i;
+
+for ( i = 0; i < 100; i++ ) {
+    x = ( randu() * 100.0 ) - 50.0;
+    console.log( '2^%d = %d', x, exp2f( x ) );
+}
````
@aayush0325 We can use `@stdlib/random/array/uniform` here.

gunjjoshi replied:
Same can be followed in `examples/index.js` too.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,63 @@
+
+// VARIABLES //
+
+var exp2 = tryRequire( resolve( __dirname, './../lib/native.js' ) );
```
```suggestion
var exp2f = tryRequire( resolve( __dirname, './../lib/native.js' ) );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,63 @@
+
+var exp2 = tryRequire( resolve( __dirname, './../lib/native.js' ) );
+var opts = {
+	'skip': ( exp2 instanceof Error )
```
```suggestion
	'skip': ( exp2f instanceof Error )
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,63 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		y = exp2( x[ i%x.length ] );
```
```suggestion
		y = exp2f( x[ i%x.length ] );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,63 @@
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		y = exp2( x[ i%x.length ] );
+		if ( isnan( y ) ) {
```
```suggestion
		if ( isnanf( y ) ) {
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,63 @@
+		}
+	}
+	b.toc();
+	if ( isnan( y ) ) {
```
```suggestion
	if ( isnanf( y ) ) {
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,63 @@
+var resolve = require( 'path' ).resolve;
+var bench = require( '@stdlib/bench' );
+var uniform = require( '@stdlib/random/array/uniform' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
```
```suggestion
var isnanf = require( '@stdlib/math/base/assert/is-nanf' );
```

gunjjoshi committed:
Apply suggestions from code review

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/benchmark.native.js

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,135 @@
+static double benchmark( void ) {
+	double elapsed;
+	float x[ 10 ];
+	float y;
+	double t;
```
```suggestion
	double t;
	float y;
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,134 @@
+*
+* @return random number
+*/
+static double rand_float( void ) {
```
```suggestion
static float rand_float( void ) {
```

gunjjoshi replied:
We are returning a `float` here.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,134 @@
+static double benchmark( void ) {
+	double elapsed;
+	float x[ 10 ];
+	double y;
+	double t;
```
```suggestion
	double t;
	float y;
```

gunjjoshi replied:
`y` must be a `float`, as `exp2f` returns a `float`, in L103.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,134 @@
+	int i;
+
+	for ( i = 0; i < 10; i++ ) {
+		x[ i ] = ( 240.0 * rand_float() ) - 120.0;
```
Missing `f` suffixes.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/docs/repl.txt:
```diff
@@ -0,0 +1,29 @@
+
+{{alias}}( x )
+    Evaluates the base 2 exponential function in
+    single-precision floating-point format.
```
```suggestion
    Evaluates the base 2 exponential function in single-precision
    floating-point format.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/examples/c/example.c:
```diff
@@ -0,0 +1,33 @@
+#include <stdio.h>
+
+int main( void ) {
+    float x;
+    float v;
+    int i;
+
+    for ( i = 0; i < 100; i++ ) {
+        x = ( ( (float)rand() / (float)RAND_MAX ) * 100.0f ) - 50.0f;
+        v = stdlib_base_exp2f( x );
+        printf( "2^%f = %f\n", x, v );
+    }
```
@aayush0325 We should be using tabs here, not spaces.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/lib/main.js:
```diff
@@ -0,0 +1,127 @@
+*
+*     | arithmetic | domain      | # trials | peak    | rms     |
+*     |:----------:|:-----------:|:--------:|:-------:|:-------:|
+*     | IEEE       | -1022,+1024 | 30000    | 1.8e-16 | 5.4e-17 |
```
@aayush0325 These values for domain, number of trials, peak and rms are for the double-precision implementation, i.e., `exp2`. You'll need to check the single-precision implementation and update these.

gunjjoshi replied:
For instance, the domain for `exp2f` will be restricted to `-127,+127`, and not `-1022,+1024`.

aayush0325 replied:
https://www.netlib.org/cephes/singldoc.html#exp2 i hope this is what you're looking for!


![image](https://github.com/user-attachments/assets/3a3de061-eb83-45d0-b4df-4a222cfe1f04)

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/lib/main.js:
```diff
@@ -0,0 +1,127 @@
+	}
+	if ( x < FLOAT32_MIN_BASE2_EXPONENT ) {
+		return float64ToFloat32( 0.0 );
+	}
+	// Separate into integer and fractional parts...
```
```suggestion
	}

	// Separate into integer and fractional parts...
```

gunjjoshi replied:
Empty line before a comment.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/exp2f/lib/main.js:
```diff
@@ -0,0 +1,127 @@
+		return PINF;
+	}
+	if ( x < FLOAT32_MIN_BASE2_EXPONENT ) {
+		return float64ToFloat32( 0.0 );
```
```suggestion
		return 0.0;
```

gunjjoshi replied:
`0.0` can be returned directly here.

gunjjoshi replied:
@aayush0325 After L110, you'll need to add a conditional statement to return `1.0` if x is `0.0`. You might want to refer to the single-precision implementation (`exp2f`) from cephes, as it is a bit different from porting `exp2` directly to single-precision.

aayush0325 replied:
https://github.com/jeremybarnes/cephes/blob/master/single/exp2f.c


using this as a reference here
![image](https://github.com/user-attachments/assets/4f3da97c-f103-4646-a595-776b195b3667)

kgryte commented:
@aayush0325 You still interested in pushing this PR forward?

aayush0325 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into exp2f

aayush0325 committed:
docs: update examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
feat: update C benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: update accuracy table

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
feat: add case for x == 0.0

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: fix linting error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented:
good morning @gunjjoshi! sorry for the delay, I've linked to all the references for your ease, kindly let me know if i have to make any other changes

kgryte commented:
@anandkaranubc Would you mind taking over this PR?

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into exp2f

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/exp2f/lib/main.js:
```diff
@@ -0,0 +1,132 @@
+
+'use strict';
+
+// TODO: replace with TOMS (Openlibm) algo (updating license header and long comment)
```
@kgryte Does this comment (originally present in `exp2`) mean following https://github.com/freebsd/freebsd-src/blob/main/lib/msun/src/s_exp2f.c implementation instead?

kgryte replied:
Yes, I believe so. The original implementation looks to be based on Cephes, and, generally, we have preferred to use libm over Cephes given that libm is more actively maintained.

anandkaranubc commented:
/stdlib help

anandkaranubc commented:
/stdlib update-copyright-years

anandkaranubc reviewed (approved):
I think we can merge this now. Maybe if I can get a verification pass on if the License headers and everything around that is correct, that'd be great.

cc: @kgryte

anandkaranubc commented:
Also once this is merged, I will also migrate the `exp2` implementation wrt to FreeBSD.

kgryte commented on lib/node_modules/@stdlib/math/base/special/exp2f/package.json:
```diff
@@ -0,0 +1,70 @@
+    "doc": "./docs",
+    "example": "./examples",
+    "include": "./include",
+    "lib": "./lib",
```
```suggestion
    "lib": "./lib",
    "scripts": "./scripts",
```

kgryte reviewed (approved):
LGTM
